python/ray/data/_internal/execution/operators/input_data_buffer.py

```diff
@@ -14,33 +14,65 @@ class InputDataBuffer(PhysicalOperator):
     the arguments for read tasks.
     """
 
-    def __init__(self, input_data: List[RefBundle]):
+    def __init__(
+        self,
+        input_data: Optional[List[RefBundle]] = None,
+        input_data_factory: Callable[[], List[RefBundle]] = None,
+    ):
         """Create an InputDataBuffer.
 
         Args:
             input_data: The list of bundles to output from this operator.
+            input_data_factory: The factory to get input data, if input_data is None.
         """
-        self._input_data = input_data
-        self._num_outputs = len(input_data)
-        block_metadata = []
-        for bundle in input_data:
-            block_metadata.extend([m for (_, m) in bundle.blocks])
-        self._stats = {
-            "input": block_metadata,
-        }
+        if input_data is not None:
+            assert input_data_factory is None
+            self._input_data = input_data
+            self._is_input_initialized = True
+            self._initialize_metadata()
+        else:
+            # Initialize input lazily when execution is started.
+            assert input_data_factory is not None
+            self._input_data_factory = input_data_factory
+            self._is_input_initialized = False
         super().__init__("Input", [])
 
     def has_next(self) -> bool:
+        if not self._is_input_initialized:
+            self._initialize_input()
         return len(self._input_data) > 0
 
     def get_next(self) -> RefBundle:
+        if not self._is_input_initialized:
+            self._initialize_input()
         return self._input_data.pop(0)
 
     def num_outputs_total(self) -> Optional[int]:
+        if not self._is_input_initialized:
```

Comment:
Add a TODO to move init into the start() method that's getting added for physical operators in the other PR.

Reply:
@ericl - sure, added.